userspace/libsinsp/mpsc_priority_queue.h

```diff
@@ -90,22 +92,64 @@ class mpsc_priority_queue
 	 * element is not popped from the queue and this method returns false.
 	 */
 	template <typename Callable>
-	inline bool pop_if(const Callable& cl, OUT Elm& res)
+	inline bool try_pop_if(Elm& res, const Callable& pred)
 	{
-		elm_ptr top = m_queue_top.load();
-		if (top == nullptr || !cl(top))
+		// we check that the queue is not empty before acquiring the lock
+		if (m_queue_top == nullptr)
 		{
 			return false;
 		}
 
-		// at this point, we have a guarantee
-		// that queue.top() has priority not less than the local top,
-		// and we can pop the queue top safely
-		std::scoped_lock<Mtx> lk(m_mtx);
-		res = std::move(m_queue.top().elm);
-		m_queue.pop();
-		m_queue_top = m_queue.empty() ? nullptr : m_queue.top().elm.get();
-		return true;
+		while (true)
+		{
+			// we need to evaluate the top element against the predicate, but
+			// we must be careful in case other producers push a new element in
+			// the queue, which can potentially have more priority than the one
+			// we just checked.
+			elm_ptr top = m_queue_top.load();
+			auto should_pop = pred(*top);
+			
+			// we must not pop the element
+			if (!should_pop)
+			{
+				// check that the top-priority element ha not changed since
+				// we evaluated it, otherwise keep looping
+				if (top == m_queue_top.load())
+				{
+					return false;
+				}
+				continue;
+			}
+
+			// check that the top-priority elem has changed since evaluating it,
+			// otherwise keep looping. We check this before acquiring the lock
+			// as an extra concurrency optimization.
+			if (top != m_queue_top.load())
```

Comment:
not sure if we have a real optimization here, most of the time we perform a useless check while probably only a few times it will save us from locking, it's also true that it is just an `if`...

Reply:
Assuming there has been no pusher in between, I would expect the volatile value of m_queue_top to not be invalidated in the cache due to some cache coherence strategy, and thus the atomic read to be less costly (happy to be proved wrong if any real expert will read this). As such, this check will definitely be an optimization in case we'll ever have a burst of pushers. In such a case, attempting acquiring the lock while pushers are trying to do the same would just make it worse.

Reply:
yes my point here is that the cases in which this extra check saves us from acquiring the lock should be way less frequent than the times in which we perform this check without having a real benefit. Moreover, after this check, there is still the possibility that the m_queue_top is invalidated before acquiring the lock...but yes probably just personal taste, let's keep it as it is